src/types.rs

```diff
@@ -185,24 +183,28 @@ impl BoundedStorable for StringStorable {
     const IS_FIXED_SIZE: bool = false;
 }
 
-impl Storable for PrincipalStorable {
-    fn to_bytes(&self) -> Cow<[u8]> {
-        Cow::from(self.0.as_slice())
+#[derive(Clone)]
+pub struct ApiKey(String);
+
+impl ApiKey {
+    pub fn empty() -> ApiKey {
```

Comment:
IMO should not be needed. `get_api_key` should return an `Option` since it is indeed the case that not all providers have an API key.

Reply:
This definitely makes more sense now that we have `public_url`. Updated.